libkbfs/block_journal.go

```diff
@@ -992,62 +663,37 @@ func (j *blockJournal) removeFlushedEntry(ctx context.Context,
 			ordinal, earliestOrdinal)
 	}
 
-	// Fix up the block byte count if we've finished a Put.
-	if entry.Op == blockPutOp && !entry.Ignore {
-		id, _, err := entry.getSingleContext()
-		if err != nil {
-			return 0, err
-		}
-		flushedBytes, err = j.getDataSize(id)
-		if err != nil {
-			return 0, err
-		}
-
-		if flushedBytes > j.unflushedBytes {
-			return 0, fmt.Errorf("Block %v is bigger than our current count "+
-				"of journal block bytes (%d > %d)", id,
-				flushedBytes, j.unflushedBytes)
-		}
-		j.unflushedBytes -= flushedBytes
-	}
-
 	_, err = j.j.removeEarliest()
 	if err != nil {
 		return 0, err
 	}
 
 	// Remove any of the entry's refs that hasn't been modified by
 	// a subsequent block op (i.e., that has earliestOrdinal as a
-	// tag).
+	// tag). Has no effect for removeRefsOp (since those are
+	// already removed) or mdRevMarkerOp (which has no
+	// references).
 	for id, idContexts := range entry.Contexts {
-		refs := j.refs[id]
-		if len(refs) == 0 {
-			continue
+		liveCount, err := j.s.removeReferences(
+			id, idContexts, earliestOrdinal.String())
+		if err != nil {
+			return 0, err
 		}
-		for _, context := range idContexts {
-			err := refs.remove(context, earliestOrdinal)
+		// If j.saveUntilMDFlush is non-nil, then postpone
+		// garbage collection until it becomes nil.
+		if j.saveUntilMDFlush == nil && liveCount == 0 {
+			// Garbage-collect the old entry if we are not
+			// saving blocks until the next MD flush.
+			flushedBytes, err := j.remove(id)
 			if err != nil {
 				return 0, err
 			}
-			if len(refs) == 0 {
-				delete(j.refs, id)
-
-				// Garbage-collect the old entry if we are not saving
-				// blocks until the next MD flush.  TODO: we'll
-				// eventually need a sweeper to clean up entries left
-				// behind if we crash here.
-				if j.saveUntilMDFlush == nil {
-					err = j.removeBlockData(id)
-					if err != nil {
-						return 0, err
-					}
-				}
-				break
-			}
+
+			totalFlushedBytes += flushedBytes
```

Comment:
If we want to avoid file reads, it might be simpler to touch/rm a flag file instead. The flag file might store info, but it could be such that if it's present, we know that it's unflushed, so the walk can just check for the existence of that file.

There's a bit of an edge case regarding multiple puts for the same block, in which I think we want to count that block twice (or n times). But that should be rare, so it should be possible to do something where it'll only do a file read if there are multiple puts.